16.0 reconversion

stephansainleger opened the pull request.

Commit:
[IMP] lcc_lokavaluto_app_connection: create debit.request model

Commit:
[IMP] lcc_lokavaluto_app_connection: add commission invoices

for reconversion process

Commit:
[FIX] lcc_lokavaluto_app_connection: fix request states calculation

- refactoring of the _invoice_paid_hook() function (account.move)
- change _post() into action_post() as function to override for the
- debit request state computation
- call to debit.request compute_state() function when invoices are created
- fix of the company parameter retriveval on credit.request write() function

Commit:
[IMP] lcc_lokavaluto_app_connection: add reconversion checking cron

This cron will launch a checking process to know if some reconversion
transactions has bee received by the safe wallets of the digital
currencies backend.
Some functions are empty and must be overiden in currency backend dedicated add-ons.

Commit:
[IMP] lcc_comchain_base: add reconversion checking process

Commit:
[IMP] lcc_cyclos_base: add reconversion checking process

Commit:
[IMP] lcc_cyclos_base: add Cyclos debit wallet config

Commit:
[IMP] lcc_lokavaluto_app_connection: add reconversion rule

to indicate if reconversions are allowed for a wallet.

Commit:
[IMP] lcc_comchain_base: add reconversion auth in wallet json data

Commit:
[IMP] lcc_cyclos_base: add reconversion auth in wallet json data

Commit:
[IMP] lcc_lokavaluto_app_connection: add reconversions API service

to return the reconversion status for the transactions requested.

Commit:
[IMP] lcc_lokavaluto_app_connection: refactor digital currency menu

create a new top menu for an app "Digital Currencies", with all the
menuitem available below.

Commit:
[CLN] global: rename res.partner.backend files in wallet

one of the cleaning commit to recover the actual concept of the
res.partner.backend: there are wallets, and not digital currency backends

Commit:
[FIX] lcc_lokavaluto_app_connection: add missing security rule

which allow access to the wizard create.credit.request (not needed in
v12, forgotten in the migration to v16)

Commit:
[IMP] lcc_lokavaluto_app_connection: move button in wallet form view

header

Commit:
[IMP] lcc_wallet_qr: move button in wallet form view header

Commit:
[I18N] lcc_lokavaluto_app_connection: update french translations

Commit:
[I18N] lcc_comchain_base: update french translations

Commit:
[I18N] lcc_cyclos_base: update french translations

Commit:
[IMP] lcc_comchain_base: refactor comchain settings view

Commit:
[IMP] lcc_members: remove user types display in user form

functionnality move to new add-on user_types_displayed_in_user_form
https://github.com/elabore-coop/ux-tools/tree/16.0/user_types_displayed_in_user_form

Commit:
[FIX] lcc_lokavaluto_app_connection: corrects typo

vaab commented on lcc_lokavaluto_app_connection/models/account_invoice.py:
```diff
@@ -13,6 +13,9 @@ class AccountInvoice(models.Model):
         "credit.request", "invoice_id", string="Credit Requests"
     )
+    debit_request_ids = fields.One2many(
+        "debit.request", "debit_move_id", string="Credit Requests"
```
typo: the description is wrong

stephansainleger replied:
Fixed in an following commit

vaab commented on lcc_lokavaluto_app_connection/models/account_invoice.py:
```diff
@@ -60,6 +69,19 @@ def _compute_has_numeric_lcc_products(self):
 
+    def _compute_debit_request_ids(self):
+        for move in self:
+            move.debit_request_ids = self.env["debit.request"].search(["|", ("debit_move_id", "=", move.id), ("commission_move_id", "=", move.id)])
```
there are no commission_move_id yet

vaab commented on lcc_lokavaluto_app_connection/models/account_invoice.py:
```diff
@@ -60,6 +69,19 @@ def _compute_has_numeric_lcc_products(self):
+            self.digital_currency_invoice_type = "credit"
+        else:
+            self.digital_currency_invoice_type = "none"
+
```
so if there are both debit request and credit requests, it is considered a debit ... The only place where it is used in current code could do without this (`account_invoice.xml`) using `('credit_request_ids', '=', '[]')`.

stephansainleger replied:
fixed in https://github.com/Lokavaluto/lokavaluto-addons/pull/301/commits/16d9cddff8c86287d6c4ce3cade8ac8d266a1550

vaab commented on lcc_lokavaluto_app_connection/models/account_invoice.py:
```diff
@@ -69,8 +91,23 @@ def _invoice_paid_hook(self):
                     continue
                 # Set the state in "pending" to launch the top up process
                 request.write({"state": "pending"})
+        for invoice in self.filtered(lambda move: move.is_invoice() and move.is_purchase_document() and move.has_numeric_lcc_products):
```
Could factorize with previous loop, and have one if on : is it a sale or a purchase ? It'll be more efficient

stephansainleger replied:
Fixed in an following commit

vaab commented on lcc_lokavaluto_app_connection/models/debit_request.py:
```diff
@@ -0,0 +1,130 @@
+            raise UserError("The wallet is missing.")
+        if not self.transaction_id and raise_error:
+            raise UserError("The transaction ID is missing.")
+        return (self.amount > 0) and self.wallet_id and self.transaction_id
```
You are duplicating the tests for no reasons... You could raise always errors in this function, and call it with try/except when you don't want the exceptions.

stephansainleger replied:
solved, raise_error attribute removed, usage of try/except when needed.

vaab commented on lcc_lokavaluto_app_connection/models/debit_request.py:
```diff
@@ -0,0 +1,130 @@
+                )
+
+            if request.debit_move_id and request.debit_move_id.state == "draft":
+                request.debit_move_id.unlink()
```
you duplicate some tests, and some lines seem unreachable (can you have a state cancelled or paid with a connected non-draft invoice ?). In some case even, it seem it seem it could accept unlink with a `debit_move_id` (but not draft) connected that it'll leave untouched. Currently, I'm not sure if all this is done on purpose.

stephansainleger replied:
fixed

vaab commented on lcc_lokavaluto_app_connection/models/debit_request.py:
```diff
@@ -0,0 +1,130 @@
+            if request.debit_move_state == "paid":
+                request.state = "paid"
+            if request.debit_move_id.state == "cancelled":
+                request.state = "cancelled"
```
these could be `elif`

stephansainleger replied:
fixed in a following commit

vaab commented on lcc_lokavaluto_app_connection/models/debit_request.py:
```diff
@@ -0,0 +1,130 @@
+    ## INVOICE CREATION PROCESSES
+    #############################
+    
+    def create_invoices(self):
+        self.create_debit_invoices()
```
why this indirection ?

stephansainleger replied:
completed later with another function... it is a Back To The Future design ^^'

vaab reviewed (commented):
First commit comments...

vaab commented on lcc_lokavaluto_app_connection/models/account_invoice.py:
```diff
@@ -13,8 +13,8 @@ class AccountInvoice(models.Model):
-    debit_request_ids = fields.One2many(
-        "debit.request", "debit_move_id", string="Credit Requests"
+    debit_request_ids = fields.Many2many(
+        "debit.request", compute="_compute_debit_request_ids", string="Debit Requests"
```
Shouldn't it still stay a One2Many ?

stephansainleger replied:
If my understanding is good, if `debit_request_ids` is a one2many, this means that there should be a many2one field that link the debit request to the invoice.
the issue is that here we have 2 many2one that must be checked : `debit_move_id` and `commission_move_id` : according to the type of invoice (sale or purchase), we should consider one or the other.

vaab commented on lcc_lokavaluto_app_connection/models/company.py:
```diff
@@ -1,4 +1,4 @@
-from odoo import models, fields
+from odoo import models, fields, api
```
not clear why this import is needed in this commit

stephansainleger replied:
removed

vaab commented on lcc_lokavaluto_app_connection/models/debit_request.py:
```diff
@@ -75,15 +86,44 @@ def unlink(self):
+            if request.debit_move_state == "cancelled":
+                # Debit move cancellation means there is no debit to perform, then all the process is cancelled.
                 request.state = "cancelled"
+                continue
+
+            if request.commission_move_state == "cancelled":
+                # If the commission invoice is cancelled, we do not consider it anymore for the request status.
+                request.state = self._convert_status(request.debit_move_state)
+                continue
+
+            if request.debit_move_state == "draft" or request.commission_move_state == "draft":
+                # If any of the invoices is in Draft, the request is still considered as "received"
+                request.state = self._convert_status("draft")
+                continue
+
+            if request.debit_move_state == "posted" or request.commission_move_state == "posted":
+                request.state = self._convert_status("posted")
+                continue
+
+            if request.debit_move_state == "paid" or request.commission_move_state == "paid":
+                request.state = self._convert_status("paid")
```
I'm a little lost into the intent here, and fear that it's a little too brittle.

vaab commented on lcc_lokavaluto_app_connection/models/debit_request.py:
```diff
@@ -127,4 +168,33 @@ def _get_debit_invoice_line_values(self):
+
+    def create_commission_invoices(self):
+        for request in self:
+            self.is_ready_to_invoice(raise_error=True)
```
this test is done twice

vaab commented on lcc_lokavaluto_app_connection/models/res_partner_backend.py:
```diff
@@ -1,8 +1,10 @@
+from odoo.tools.safe_eval import safe_eval
 import logging
 
+
+
```
unecessary, some pass with ruff in incrementally can be applied at the end if you wish

vaab commented on lcc_lokavaluto_app_connection/models/res_partner_backend.py:
```diff
@@ -95,4 +97,16 @@ def get_wallet_balance(self):
+        rules = self.env["commission.rule"].search([("active", "=", True)])
+        for rule in rules:
+            # Get all the wallet matching the rule
+            wallets = self.search(safe_eval(rule.wallet_domain))
```
I'm not sure to see how, without providing a context, this `safe_eval` is useful...

stephansainleger replied:
well I don't know how it works exactly, but without the safe_eval() it doesn't work ^^'

vaab commented on lcc_lokavaluto_app_connection/models/res_partner_backend.py:
```diff
@@ -95,4 +97,16 @@ def get_wallet_balance(self):
+            # Check if current wallet (self) is in the matching wallets
+            if wallets.filtered(lambda x: x.id == self.id):
+                # First rule matched is returned
+                return rule
```
only the first rule will match... first in which order ? the default order as you didn't specify any `_order`. Is that what you want ? This computation will also be quite costly written like that. You should add the `('id', '=', self.id)` to the search at least and avoid doing the test in python.

stephansainleger replied:
the idea is that as soon as one rule match, we consider that the reconversion is allowed, whatever the rule is.
I have modified the code to make it less costly.

vaab reviewed (commented):
Review of second commit

vaab commented on lcc_lokavaluto_app_connection/models/account_invoice.py:
```diff
@@ -84,33 +84,42 @@ def _compute_digital_currency_invoice_type(self):
-                request.write({"debit_move_state": "paid"})
+                if invoice.is_sale_document():
+                    state_value = {"commission_move_state": "paid"}
+                if invoice.is_purchase_document():
```
elif

vaab commented on lcc_lokavaluto_app_connection/models/account_invoice.py:
```diff
@@ -84,33 +84,42 @@ def _compute_digital_currency_invoice_type(self):
-                request.write({"debit_move_state": "posted"})
+                if invoice.is_sale_document():
+                    state_value = {"commission_move_state": "posted"}
+                if invoice.is_purchase_document():
```
elif

vaab commented on lcc_lokavaluto_app_connection/models/account_invoice.py:
```diff
@@ -84,33 +84,42 @@ def _compute_digital_currency_invoice_type(self):
+                    state_value = {"commission_move_state": "paid"}
+                if invoice.is_purchase_document():
+                    state_value = {"debit_move_state": "paid"}
+                request.write(state_value)
```
if none of the 2 condition above matches, it can fail with an error or worse, write the state of the last element seen

vaab commented on lcc_lokavaluto_app_connection/models/account_invoice.py:
```diff
@@ -84,33 +84,42 @@ def _compute_digital_currency_invoice_type(self):
+                    state_value = {"commission_move_state": "posted"}
+                if invoice.is_purchase_document():
+                    state_value = {"debit_move_state": "posted"}
+                request.write(state_value)
```
if none of the 2 condition above matches, it can fail with an error or worse, write the state of the last element seen.
Note that if you are sure that if it can't happen, then make an if/else rather than 2 ifs... and best if you have any doubt, is to raise an clear error.

vaab commented on lcc_lokavaluto_app_connection/models/credit_request.py:
```diff
@@ -58,7 +58,7 @@ def write(self, vals):
             if request.state == "pending":
                 # The top up has been paid, the credit process can start
-                if request.partner_id.company_id.activate_automatic_topup:
+                if self.env.user.company_id.activate_automatic_topup:
```
are we sure to want to check who is actively calling this `write` ? Not real sure to understand why this change.

vaab reviewed (commented):
Several typos in the commit message of the third commit.

vaab commented on lcc_lokavaluto_app_connection/models/company.py:
```diff
@@ -9,3 +11,57 @@ class Company(models.Model):
     activate_automatic_topup = fields.Boolean("Activate Automatic Topup")
     commission_product_id = fields.Many2one("product.product", string="Commission Product")
+
+
```
pep8 requires only one blank line here ;-)

vaab commented on lcc_lokavaluto_app_connection/models/company.py:
```diff
@@ -9,3 +11,57 @@ class Company(models.Model):
+
+
+    def _cron_import_new_digital_currency_debit_requests(self):
+        """Create the debit requests in Odoo for all the debit
+        transactions performed since the last check."""
```
pep 257 strongly recommend one-line summary, and if multi-line, one-line summary, blank line then any content.

vaab commented on lcc_lokavaluto_app_connection/models/company.py:
```diff
@@ -9,3 +11,57 @@ class Company(models.Model):
+
+    def _retrieve_last_debit_transactions(self):
+        """TO OVERIDE in digital currency backend dedicated add-ons
+        A list of transactions (dictionnary) is expected, with the following data:
```
dictionary (only one n).

Commit:
[MIG] pkg: make 'demo, test, dev1, dev2, dev3' branches deploy code

vaab commented on lcc_lokavaluto_app_connection/models/company.py:
```diff
@@ -9,3 +11,57 @@ class Company(models.Model):
+        # - if yes, do nothing
+        # - if no, create a new debit request
+        for transaction in transactions:
+            tx_id = transaction.get("tx_id", "")
```
hmm, why silence this error in the data ? A log at least should be important. And we can probably `continue` also.

stephansainleger replied:
add a check to directly continue before performing the search on debit.requests in the following line.
add warning log too.

vaab commented on lcc_lokavaluto_app_connection/models/company.py:
```diff
@@ -9,3 +11,57 @@ class Company(models.Model):
+
+
+    def _build_debit_request_values(self, transaction):
+        wallet_ids = self.env["res.partner.backend"].search([("name", "=", transaction.get("sender"))])
```
have you made sure that the names have a constraint of unicity on wallets ?

stephansainleger replied:
no unicity constraint for now, but we log an error if several wallets are found.

vaab commented on lcc_lokavaluto_app_connection/models/company.py:
```diff
@@ -9,3 +11,57 @@ class Company(models.Model):
+            "wallet_id" : wallet_ids[0].id,
+            "amount" : transaction.get("amount"),
+            "transaction_id" : transaction.get("tx_id"),
+            "transaction_timestamp": transaction.get("tx_timestamp"),
```
I'm concerned that you didn't check if any of these values are actually set. As it is the responsability of a plugin, it could be nice to check the content and make a nice helpful message of what was forgotten or what went wrong instead of writing bogus values in database and wait for a crash later one.

stephansainleger replied:
this values are filled by function retrieve_last_debit_transactions() defined in lcc_comchain_base and lcc_cyclos_base add-ons. This function check that these values exist.

vaab reviewed (commented):
a typo (overridden : 2 r, 2 d) and strange wrapping in comment of commit number 4 ! ;-)

vaab commented on lcc_comchain_base/models/company.py:
```diff
@@ -19,3 +26,54 @@ def _default_messages(self):
+        except Exception as e:
+            _logger.error(tools.format_last_exception())
+
+        return res
```
res might be incompletely filled with some transaction of the current block. You should probably reraise the exception ? You probably have a good reason why you didn't.

stephansainleger replied:
fixed: we use a tmp list `comchain_res` to fill the retrieved transaction, and this `comchain_res` is added to `res` only with all the process went smoothly.

vaab commented on lcc_comchain_base/models/company.py:
```diff
@@ -19,3 +26,54 @@ def _default_messages(self):
+                        continue # this transaction is not a digital currency transfer
+                    if not full_tx.direction == 1:
+                        continue # this transaction is not a reconversion
+                    _logger.info("Transactions %s retrieved" % tx.hash)
```
typo: transaction~s~

vaab commented on lcc_comchain_base/models/company.py:
```diff
@@ -19,3 +26,54 @@ def _default_messages(self):
+                        continue # this transaction is not a reconversion
+                    _logger.info("Transactions %s retrieved" % tx.hash)
+                    sender_adress = full_tx.add2.lstrip("0x")
+                    matched_wallet = self.env["res.partner.backend"].search([("comchain_id", "ilike", sender_adress)])
```
why `ilike` ?

vaab commented on lcc_comchain_base/models/company.py:
```diff
@@ -19,3 +26,54 @@ def _default_messages(self):
+                    if not full_tx.direction == 1:
+                        continue # this transaction is not a reconversion
+                    _logger.info("Transactions %s retrieved" % tx.hash)
+                    sender_adress = full_tx.add2.lstrip("0x")
```
typo: address (with 2 d)

vaab commented on lcc_comchain_base/models/company.py:
```diff
@@ -19,3 +26,54 @@ def _default_messages(self):
+                    sender_adress = full_tx.add2.lstrip("0x")
+                    matched_wallet = self.env["res.partner.backend"].search([("comchain_id", "ilike", sender_adress)])
+                    if len(matched_wallet) == 0:
+                        continue # this transactions comes from a wallet no registered in current Odoo database
```
typo: `not registered` and this transaction~s~

vaab commented on lcc_comchain_base/models/company.py:
```diff
@@ -19,3 +26,54 @@ def _default_messages(self):
+                    string_amount = str(full_tx.sent)
+                    res.append({
+                        "sender": "comchain:%s" % sender_adress,
+                        "amount": f"{string_amount[0:-2]}.{string_amount[-2:]}",
```
Warning here, this is not guaranteed to work, if full_tx.sent = '1' you are screwed ;-)... you need to `.zfill(2)`

vaab commented on lcc_comchain_base/models/company.py:
```diff
@@ -19,3 +26,54 @@ def _default_messages(self):
+                        "sender": "comchain:%s" % sender_adress,
+                        "amount": f"{string_amount[0:-2]}.{string_amount[-2:]}",
+                        "tx_id": tx.hash,
+                        "tx_timestamp": datetime.fromtimestamp(int(full_tx.receivedat["value"])),
```
This value doesn't mean anything. If your version of pyc3l supports it, try `full_tx.received_at` (which is a datetime already). Otherwise you'll also probably have a wrong datetime (it is an utc timestamp, and you should store timezone aware datetime...)

stephansainleger replied:
fixed by using directly full_tx.received_at

vaab commented on lcc_comchain_base/models/company.py:
```diff
@@ -19,3 +26,54 @@ def _default_messages(self):
+                    })
+
+            company_id.last_block_checked_nb = last_block_id
+        except Exception as e:
+            _logger.error(tools.format_last_exception())
```
For me that's probably a bad idea. There are many exception that are meant to break totally the flow of odoo python process. My problem is that you are using a 'except Exception' which is too wide, OR that you don't re-raise it.

stephansainleger replied:
As I am clearly not a error managment expert yet, I will need your help on that point :-)

vaab reviewed (changes requested):
some important issue here.

vaab committed:
[FIX] lcc_cyclos_base: prevent displaying the cyclos account creation button to unauthorized user

Will avoid displaying the button to people outside the group.

Commit:
[FIX] lcc_comchain_base: remove potential binary error

for some numbers, the multiplication can lead to an wrong result due
to binary issues. The add of function round() for the cents
calculation will prevent any side effect.

Commit:
[FIX] lcc_lokavaluto_app_connection: remove failed requests from

Commit:
[FIX] lcc_comchain_base: improve error management on pyc3l requests

Commit:
[FIX] lcc_comchain_base: manage error case when transaction data is empty

Commit:
[IMP] lcc_lokavaluto_app_connection: change credit/debit pages display rules

vaab committed:
[CLN] apply ruff on all files

Commit:
[FIX] lcc_comchain_base: change company  retrieval in comchain wallet credit

Commit:
[FIX] lcc_comchain/cyclos_base: add type check in get_wallet_balance

vaab committed:
[FIX] lcc_lokavaluto_app_connection/services/auth_services: correct ``can_reset_password`` from ``GET`` to ``POST``

Monujo is making POST requests here for now.

vaab committed:
[FIX] lcc_members/controllers/auth_signup: prevent exception about ``lang`` missing

Probably an odoo 16.0 change of request object interface.

vaab committed:
[FIX] lcc_{comchain,cyclos}_base: remove dangerous ``try/except``

vaab committed:
[IMP] lcc_comchain_base: ensure that the block range makes sense

vaab committed:
[FIX] lcc_{lokavaluto_app_connection,{comchain,cyclos}_base}: strenghten reconversion collection

- use generator for on-the-fly validation and commit of database
  transaction
- raise exceptions whenever possible as soon as possible
- fix UTC dates generation to avoid smashing timezones
- guard conversion to floats against loss of accuracy
- rename field tx_* -> transaction_* to match the database model and
- rename field transaction_timestamp -> transaction_date ... because
  it's a date and not a timestamp
- reduce code overhead

vaab committed:
[FIX] lcc_{lokavaluto_app_connection,{comchain,cyclos}_base}: prevent safe partners to appears as possible transfer recipients

vaab committed:
[FIX] lcc_{comchain,cyclos}_base: allow safe wallets to be advertised to monujo

vaab committed:
[FIX] lcc_lokavaluto_app_connection: add ``sale_management`` to dependencies

Commit:
[IMP] lcc_lokavaluto_app_connection: prevent wallet delete with security rules

Commit:
[FIX] lcc_lokavaluto_app_connection: typo

Commit:
[CLN] all: add missing licence

Commit:
[IMP] lcc_lokavaluto_app_connection: add missing model descriptions

Commit:
[IMP] lcc_lokavaluto_app_connection: new links between account move and debit requests

Commit:
[IMP] lcc_lokavaluto_app_connection: add sequence field in model commission_rule

Commit:
[IMP] lcc_lokavaluto_app_connection: simplify get_wallet_commission_rule()

vaab committed:
[FIX] lcc_{comchain,cyclos}_base: don't provide the safe wallet if reconversion is not allowed

LaetitiaElabore committed:
[FIX]lcc_lokavaluto_app_connection:fix bug name is required at company creation

LaetitiaElabore committed:
[FIX]lcc_lokavaluto_app_connection:fix bug on partner at invoice creation
